File: largest_prime_facter.rb
#!/usr/bin/env ruby


# Problem 3 from projecteuler.net
#
# The prime factors of 13195 are 5, 7, 13 and 29.
#
# What is the largest prime factor of the number 600851475143 ?

def list_of_primes num_to_check
  orig_num = num_to_check.to_i
  num_to_check = (num_to_check.to_i / 2).floor
  #num_to_check = Math.sqrt(num_to_check.to_i).floor
  puts "Checking up to #{num_to_check}"
  a = ['x','x']
  return_list = []
  2.upto(num_to_check) do |j|
    next if a[j] == 'x'
    #puts "Working on #{j}"
    2.upto(num_to_check) do |i|
      break if i * j > num_to_check
      a[i * j] = 'x'
    end
    #puts "At #{j}, array has #{a.size} elements"
  end

  a.each_with_index do |n,i|
    break if i > num_to_check
    if n != 'x'
      return_list << i
    end
  end

  return_list
end

if __FILE__ == $0

  num = ARGV[0].to_i
  
  primes = list_of_primes(num)
  
  primes.each do |p|
    printf "%i ", p if ( num % p.to_i == 0 )
  end
end
# vim: ai ts=2 sts=2 et sw=2 ft=ruby

File: largest_prime_facter.py
#!/usr/bin/env python


# Problem 3 from projecteuler.net
#
# The prime factors of 13195 are 5, 7, 13 and 29.
#
# What is the largest prime factor of the number 600851475143 ?

import math

target_num = 600851475143
target_num = 13195
prime_factors = []

def is_prime(num):
  #print("Checking %i" % num)
  max = int(math.ceil(math.sqrt(num)))
  #print("max = %i" % max)
  for n in range(2, max):
    #print("Checking %i" % n)
    if num % n == 0:
      return False
  return True

def is_a_factor(big_num, little_num):
  if big_num % little_num == 0:
    return True

for i in range(3,target_num):
  if i % 2 == 0:
    continue
  if not is_a_factor(target_num, i):
    continue
  if (not is_prime(i)):
    continue
  prime_factors.append(i)
  print("%i" % i)
